spec/unit/provider/user/windows_adsi_spec.rb

```diff
@@ -60,6 +60,88 @@
     end
   end
 
+  describe "groups_insync?" do
+
+    let(:group1) { stub(:account => 'group1', :domain => '.', :to_s => 'group1sid') }
+    let(:group2) { stub(:account => 'group2', :domain => '.', :to_s => 'group2sid') }
+    let(:group3) { stub(:account => 'group3', :domain => '.', :to_s => 'group3sid') }
+
+    before :each do
+      Puppet::Util::Windows::SID.stubs(:name_to_sid_object).with('group1').returns(group1)
+      Puppet::Util::Windows::SID.stubs(:name_to_sid_object).with('group2').returns(group2)
+      Puppet::Util::Windows::SID.stubs(:name_to_sid_object).with('group3').returns(group3)
+    end
+
+    it "should return false when current is nil" do
+      expect(provider.groups_insync?(nil, ['group2'])).to be_falsey
```

Comment:
The output is pretty cool.

```
 1) Puppet::Type::Group::ProviderWindows_adsi group type :members property helpers #members_insync? should return true when current user(s) and should user(s) are nil lists
     Failure/Error: expect(provider.members_insync?(nil, nil)).to be_truthy
       expected: truthy value
            got: false
     # ./spec/unit/provider/group/windows_adsi_spec.rb:63:in `block (4 levels) in <top (required)>'

  2) Puppet::Type::User::ProviderWindows_adsi groups_insync? when membership => inclusive should return true when should is nil
     Failure/Error: expect(provider).to be_groups_insync(['group1'], nil)
       expected `User[testuser](provider=windows_adsi).groups_insync?(["group1"], nil)` to return true, got false
     # ./spec/unit/provider/user/windows_adsi_spec.rb:97:in `block (4 levels) in <top (required)>'
```

However the code reader's experience is really really bad IMHO. I don't immediately know what it going on looking at that code and if I have similarly named methods e.g. `in_sync` vs `in_sync?`, which one does it match on? It's not immediately obvious unfortunately. My method named example isn't one we use but I have seen them from time to time. The actual test name is really helpful in determining the error, IMHO.

`Puppet::Type::Group::ProviderWindows_adsi group type :members property helpers #members_insync? should return true when current user(s) and should user(s) are nil lists` tells me immediately what is wrong regardless of the rest of the test output if the test fails.